### DISCUSSION TITLE
Why not Antd tables?

### DISCUSSION THREAD

**Discussion by vrilcode** [lowdefy/lowdefy#1468] [Ideas and feature requests]
Is there any reason why Lowdefy offers Ag-Grid integration instead of Antd tables? Ag-Grid is quite full-featured for sure, but at the moment it feels a little alien compared to other components (compared to Lists especially). And Antd tables also offer a lot of features, especially the Pro version:
* https://procomponents.ant.design/en-US/components/editable-table/
* https://procomponents.ant.design/en-US/components/drag-sort-table/
* https://procomponents.ant.design/en-US/components/table/
* https://ant.design/components/table/

So far I've not implemented any extension for Lowdefy. Would it be a lot of work to fully integrate (events, blocks inside cells, ...) Antd (pro) tables as Lowdefy v4 plugin(s)? Is it just "coupling" of Antd React components and Lowdefy interfaces or is it more work?

**Comment by Gervwyk** [comment]
Hi @vrilcode 

Ag-Grid is alien for sure, and Antd Tables would be a better fit. 
The main reason this is not implemented yet is that we have not finalized the schema for blocks inside of cells, and we try to prevent schema breaking changes at all cost. 

It's really high on our list of blocks to add, and maybe it makes sense to add a simplified version at first to cover the base use cases. 
Feature rich tables can become quite complex and it has been convenient to hand all to that off to Ag-Grid, which is excellent, however, I 100% agree that we need Antd Tables.

Depending on which features are supported, building a v4 plugin to add it can be simple. 
Would this be something you'd consider doing? Do you have some React experience? 
We'd love to help and accept a PR for this.

**Reply by vrilcode** [reply]
Hey @Gervwyk,

> Ag-Grid is alien for sure, and Antd Tables would be a better fit. The main reason this is not implemented yet is that we have not finalized the schema for blocks inside of cells, and we try to prevent schema breaking changes at all cost.

Thanks for explanation. I am new to Lowdefy, but as far as I see, tables are somehow similar to lists - at least from Lowdefy user perspective. Maybe I am wrong in terms of implementation.

> It's really high on our list of blocks to add, and maybe it makes sense to add a simplified version at first to cover the base use cases.
Feature rich tables can become quite complex and it has been convenient to hand all to that off to Ag-Grid, which is excellent, however, I 100% agree that we need Antd Tables.

I think even a simple version (as long as Buttons in cells can be used) would be a real improvement. Currently I am "misusing" Lists with Spans for fake tables (because Ag-Grid cannot be used with buttons afaik).

Tables may not always be the best way to display (or enter) data (especially for responsive/mobile design), but they are quite universal. Nearly every (repeating) data can be expressed in form of tables, very quickly. So it's a good starting point for (optional) later improvements towards more specific ways to display data. I think because of their universality, nearly every low-code tool offers them.

> Depending on which features are supported, building a v4 plugin to add it can be simple. Would this be something you'd consider doing? Do you have some React experience? We'd love to help and accept a PR for this.

Currently I'am trying to implement my first app with Lowdefy v3. So I am still learning Lowdefy. But after that and if v4 is released I would definitely try to create a simple Antd table plugin. Yes, I have some React experience. If I will do it, I would make a public repository and I will post a link here.

PS: I've considered a lot of low-code tools before and for me - as a "normal" programmer - YAML files all around, was a quite strange approach in the beginning. But I've to say, that I am really convinced by Lowdefy's architecture now, because for me it's not so important to avoid keyboard entry in general (UI is nice to start, but can be annoying for repeating tasks or revision). I think most important is to avoid ever and ever repeating coding work for simple CRUD applications. And I think there is no tool outside, which gives you the opportunity to create an app with less configuration than Lowdefy. To summarize: less typing and reusability are the most important factors for me.
